### ISSUE TITLE
PLIP: Speedup registry lookup

### ISSUE THREAD

**Issue by jensens** [plone/Products.CMFPlone#3728]
## Abstract

At present, the Plone registry is accessed multiple times during a request-response cycle. This can be optimized by implementing a subscriber that looks up the Plone registry upon traversal when entering the Plone site root, then sets it as an attribute of the request object named "request.registry". This would reduce the need to perform repeated look-ups, as the registry object can be utilized wherever a request is available, rather than having to perform the lookup every time.

## Motivation

Speed up Plone. Dependent on the page rendered this speeds up a request between 10 and 100ms (depends on complexity of the page).

## Assumptions

-/-

## Proposal & Implementation

Two steps:

## Subscriber 

Implement traversal subscriber and add registry to request. 

This is implemented in `plone.base`: 
- it defines `IPloneSiteroot`
- it already depend on `plone.registry`.
- it depends (transitive) on `zope.traversing`.

A new subscriber handler is created in a new file `handler.py`

```
def registry_handler(site, event):
    registry = queryUtility(IRegistry)
    request = getattr(site, "REQUEST", None)
    # alternatively `zope.globalrequest` could be used, but this should be enough to get the current request.
    if not request:
        # this can not happen, but not sure about test scenarios.
        return
    # the next line might need a access control wrapper to be used safely in RestrictedPython. Needs check.
    setattr(request, "registry", registry)
```

The traversal subscriber is registered:

```XML
  <subscriber
      for="plone.registry.interfaces.siteroot.IPloneSiteRoot
           zope.traversing.interfaces.IBeforeTraverseEvent"
      handler=".handlers.registry_handler"
      />
```

This can be done in the Plone 6.0.x series already since no incompatibility is added.

### Replace usage of `getUtility(IRegistry)`

In packages depending on `plone.base` the usage of `getUtility` can be replaced by (depending on context) `self.request.registry` or `content.request.registry` or alike.

In tests with a registry involved, the test setup need to ensure the registry is set to the request.

## Deliverables

Step 1: New feature version of `plone.base`.
Step 2: Lots of releases of updated plone packages.

## Risks

- Carefully check if there are security measures needed for usage RestrictedPython/ PageTemplates/ Expression.
- Addons like `lineage.registry` will break. Those then need an own subscriber replacing the root registry with the local registry. 

## Participants

- Jens Klein (proposer)

**Comment by yurj** [comment]
what about doing it using memoize_contextless in plone.api, replacing the call to getUtility(IRegistry) with the call memoized? Something like:

@memoize_contextless
def getPloneRegistry()
    return getUtility(IRegistry, api.portal.get())

and use getPloneRegistry()?

**Comment by jensens** [comment]
> what about doing it using memoize_contextless in plone.api

This PLIP is not about addons, it is about the usage in Plone core.

**Comment by mauritsvanrees** [comment]
In general +1.

I would be inclined to _not_ do this in Plone 6.0 already.   If an add-on already sets `request.registry` it may break, especially if the add-on sets this attribute to something other than we expect.

Alternative, we might add this in `plone.base`:

```
def get_registry(request):
    # in 6.0 have this code:
    return getUtility(IRegistry)
    # in 6.1 replace with this code, perhaps with getUtility as fallback:
    return request.registry
```

But if in all 6.1 code `request.registry` just works, everyone will just use that, as it is much simpler.  For keeping a package compatible with both 6.0 and 6.1 it could be an option though.

**Comment by jensens** [comment]
OK, I have no idea if any addon doing this. A quick search in collective https://github.com/search?p=1&q=org%3Acollective+%22request.registry%22&type=Code does not show any usage (except one in Pyramid code)

Another question: Do we already have a branch for 6.1 development in Coredev?

Btw., your approach does not nail it. The idea is *not* to have a get_registry  function, but to directly use `request.registry`.

**Comment by mauritsvanrees** [comment]
> OK, I have no idea if any addon doing this. A quick search in collective https://github.com/search?p=1&q=org%3Acollective+%22request.registry%22&type=Code does not show any usage (except one in Pyramid code)

Fair enough.

> Another question: Do we already have a branch for 6.1 development in Coredev?

No.  We could create it whenever we think it is needed and when its absence starts hindering new developments.  I don't think that is the case yet, but I do expect we want this sometime during the first half year of 2023.  The CI team needs some advance warning.  

> Btw., your approach does not nail it. The idea is _not_ to have a get_registry function, but to directly use `request.registry`.

True, direct attribute access makes it much more attractive.

**Comment by davisagli** [comment]
I am not sure about attaching it to the request. Then code which uses the registry but didn't need a request before now needs one. (So it could break if it was running in a scenario where there was an active ZCA site but not a global request activated with setRequest)

But that's probably mostly a theoretical concern, and it would be a good improvement in developer experience in most cases.

**Comment by jensens** [comment]
> Then code which uses the registry but didn't need a request before now needs one.

No, code w/o a request can still use the old approach. It is not about replacing the classical lookup, it is about reducing the amount of lookups. In most cases there is a request available, this is the only state we have in a request-response cycle.

**Comment by davisagli** [comment]
> No, code w/o a request can still use the old approach.

Okay, fair enough.

**Comment by MrTango** [comment]
One thought, what is the registry is quite big? Having all this in the request object might bring some problems. Would it help, having it as an attribute on the site root instead?

**Comment by davisagli** [comment]
@MrTango What problems do you anticipate? I think putting it on the site root would cause more problems: there we would need to take care that it is not persisted.

**Comment by yurj** [comment]
Why big? It is just a reference to an existing python object... Or you mean that `queryUtility(IRegistry)` will process the registry and create a big object that represent the current status of the registry? This is the base of queryUtility:

https://github.com/zopefoundation/zope.interface/blob/82c447fc3682af19780cd2a69cbbff1f2f20f826/src/zope/interface/adapter.py#L636

As @davisagli said, this is already cached. Maybe the cache, for some reason, does not work for IRegistry?
